core/unit_test/TestNumericTraits.hpp

```diff
@@ -101,8 +101,8 @@ struct TestNumericTraits {
 
   KOKKOS_FUNCTION void operator()(Infinity, int, int& e) const {
     using Kokkos::Experimental::infinity;
-    auto const inf  = infinity<T>::value;
-    auto const zero = T(0);
+    T const inf  = infinity<T>::value;
```

Comment:
T -> auto here everywhere else below.
What was the issue with that code?

Reply:
`infinity` returns the `bit_comparison_type` type which doesn't have any comparisons or math functions defined.

Reply:
Document that observable behavior change in the description of your PR

Reply:
I suppose that was already true when it was defined as float...

Reply:
I added more details to the description.

Reply:
It is late so it might be a bad idea but did you consider defining the comparison for the bit comparison type that defers to the half type comparison?
Then we might not need these changes.

Reply:
I would rather have the numeric traits return the correct type and return `static const` and switch to `static constexpr` when `bit_cast` is `constexpr`. If we can't construct a constexpr `[b]half_type`, there doesn't seem much of a benefit of having the returned implementation type be `constexpr`.

Reply:
Turns out that doesn't work very well since we need `static` variables to be zero- or constant-initialized to be useable on the device.

Reply:
dd1d3e708690523e6c73d4469bb7a483c9e47f24 introduces some comparison operators which seems to work for the contexts we test.